### DISCUSSION TITLE
"unable to parse expression" with >= replaced with � in error message.

### DISCUSSION THREAD

**Discussion by rickan54** [ImageMagick/ImageMagick#4359] [Help] [answered]
I'm trying to learn and use the -Fx option and got an error I don't understand at all so I cut and pasted an example from

http://www.imagemagick.org/script/fx.php

"This FX script prints the first 10 prime numbers:" and got a very similar error: 

magick: unable to parse expression `for(prime=2,prime<30,composite=0;for(nn=2,nn<(prime/2+1),if((prime%nn)�0,composite++,);nn++);if(composite�0,debug(prime),);prime++)' @ error/fx.c/FxGetSymbol/1831.

My own program, which is using the Perl api, is showing the same � character in place of the >= operator and I've seen it in place of && as well in the julia set example from the same page (which also produces the "unable to parse expression" error.

I'm on Ubuntu 16.04 and I pasted the examples into bash.

Info on my copy of ImageMagick:

$ magick -version
Version: ImageMagick 7.0.8-0 Q16 x86_64 2018-06-14 https://www.imagemagick.org
Copyright: © 1999-2018 ImageMagick Studio LLC
License: https://www.imagemagick.org/script/license.php
Features: Cipher DPC 
Delegates (built-in): bzlib djvu fontconfig freetype gvc jbig jng jpeg lcms lqr lzma openexr png tiff wmf x xml zlib

**Comment by snibgo** [comment]
The bash script as shown in https://www.imagemagick.org/script/fx.php works fine for me, with IM V7.1.0-4:
```
magick xc: -channel gray -fx " \
  for (prime=2, prime < 30, composite=0; \
    for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++); \
      if (composite <= 0, debug(prime), ); prime++)" null:
```
The text output is:
```
[0,0].red: prime=2
[0,0].red: prime=3
[0,0].red: prime=5
[0,0].red: prime=7
[0,0].red: prime=11
[0,0].red: prime=13
[0,0].red: prime=17
[0,0].red: prime=19
[0,0].red: prime=23
[0,0].red: prime=29
```
Perhaps in Perl some characters need to be escaped.

**Reply by rickan54** [reply]
The error text shown in my post is from bash not Perl though the error text in Perl shows the same � character. I'm using quotes in Perl of the sort where no substitution is done and almost nothing needs escaping. I'm very, very experienced with Perl and I've never seen anything like this before.

Having no experience with Magick -fx, I don't know whether it's getting � instead of the >=. ==, <=, &&, etc. or whether it's doing the replacement itself to show where it's unable to parse.

If it's not doing the substitution, which I think is the most likely case, I have no idea where the substitution is happening. In one case I past into bash and in the other I create the fx expression in the vim editor as a string for Perl.

**Reply by rickan54** [reply]
With some Googling I get the idea that somewhere in ImageMagick something is trying to construct a single character out of the pairs that are causing the problem, and it's failing so it inserts the "replacement character", � instead. I've never seen this happen with bash or Perl before that I can remember.

**Comment by snibgo** [comment]
Please add "-debug command" to the start, redirect stderr to a text file, then paste the contents of that text file.
```
magick.exe -debug command xc: -channel gray -fx " \
  for (prime=2, prime < 30, composite=0; \
    for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++); \
      if (composite <= 0, debug(prime), ); prime++)" null: 2>dbg.txt
```
For me, dbg.txt contains:
```
2021-10-12T22:28:54+00:00 0:00.003 0.000u 7.1.0 Command magick.exe[8232]: magick-cli.c/MagickImageCommand/709/Command
  Initializing "C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick.exe"
2021-10-12T22:28:54+00:00 0:00.007 0.000u 7.1.0 Command magick.exe[8232]: magick-cli.c/ProcessCommandOptions/388/Command
  at CLI arg 1 - Starting ("-debug")
2021-10-12T22:28:54+00:00 0:00.011 0.000u 7.1.0 Command magick.exe[8232]: operation.c/CLISettingOptionInfo/458/Command
  at CLI arg 1 - Setting Option: -debug "command" "(null)"
2021-10-12T22:28:54+00:00 0:00.016 0.016u 7.1.0 Command magick.exe[8232]: operation.c/CLINoImageOperator/4779/Command
  at CLI arg 3 - NoImage Operator: -read "xc:" ""
2021-10-12T22:28:54+00:00 0:00.024 0.016u 7.1.0 Command magick.exe[8232]: operation.c/CLISimpleOperatorImages/3754/Command
  at CLI arg 4 - Simple Operator: -channel "gray" "(null)"
2021-10-12T22:28:54+00:00 0:00.027 0.016u 7.1.0 Command magick.exe[8232]: operation.c/CLIListOperatorImages/3844/Command
  at CLI arg 6 - List Operator: -fx "   for (prime=2, prime < 30, composite=0;     for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++);       if (composite <= 0, debug(prime), ); prime++)" "null"
[0,0].red: prime=2
[0,0].red: prime=3
[0,0].red: prime=5
[0,0].red: prime=7
[0,0].red: prime=11
[0,0].red: prime=13
[0,0].red: prime=17
[0,0].red: prime=19
[0,0].red: prime=23
[0,0].red: prime=29
2021-10-12T22:28:54+00:00 0:00.286 0.281u 7.1.0 Command magick.exe[8232]: operation.c/CLINoImageOperator/4779/Command
  at CLI arg 8 - NoImage Operator: -write "null:" ""
2021-10-12T22:28:54+00:00 0:00.290 0.281u 7.1.0 Command magick.exe[8232]: magick-cli.c/MagickImageCommand/804/Command
  Cleanup "C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick.exe"
2021-10-12T22:28:54+00:00 0:00.291 0.281u 7.1.0 Command magick.exe[8232]: magick-cli.c/MagickImageCommand/847/Command
  Exiting "C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick.exe"
```
This shows that IM is receiving the correct characters, for me.

**Reply by rickan54** [reply]
$ cat /tmp/dbg.txt 
2021-10-12T16:35:32-06:00 0:00.020 0.000u 7.0.8 Command magick[10550]: magick-cli.c/MagickImageCommand/709/Command
  Initializing "/tmp/.mount_magickuRAtps/usr/bin/magick"
2021-10-12T16:35:32-06:00 0:00.020 0.000u 7.0.8 Command magick[10550]: magick-cli.c/ProcessCommandOptions/388/Command
  at CLI arg 1 - Starting ("-debug")
2021-10-12T16:35:32-06:00 0:00.020 0.000u 7.0.8 Command magick[10550]: operation.c/CLISettingOptionInfo/438/Command
  at CLI arg 1 - Setting Option: -debug "command" "(null)"
2021-10-12T16:35:32-06:00 0:00.020 0.000u 7.0.8 Command magick[10550]: operation.c/CLINoImageOperator/4637/Command
  at CLI arg 3 - NoImage Operator: -read "xc:" ""
2021-10-12T16:35:32-06:00 0:00.030 0.000u 7.0.8 Command magick[10550]: operation.c/CLISimpleOperatorImages/3629/Command
  at CLI arg 4 - Simple Operator: -channel "gray" "(null)"
2021-10-12T16:35:32-06:00 0:00.030 0.000u 7.0.8 Command magick[10550]: operation.c/CLIListOperatorImages/3719/Command
  at CLI arg 6 - List Operator: -fx "   for (prime=2, prime < 30, composite=0;     for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++);       if (composite <= 0, debug(prime), ); prime++)" "null"
2021-10-12T16:35:32-06:00 0:00.040 0.010u 7.0.8 Command magick[10550]: magick-cli.c/MagickImageCommand/804/Command
  Cleanup "/tmp/.mount_magickuRAtps/usr/bin/magick"
2021-10-12T16:35:32-06:00 0:00.040 0.010u 7.0.8 Command magick[10550]: magick-cli.c/MagickImageCommand/847/Command
  Exiting "/tmp/.mount_magickuRAtps/usr/bin/magick"
magick: unable to parse expression `for(prime=2,prime<30,composite=0;for(nn=2,nn<(prime/2+1),if((prime%nn)�0,composite++,);nn++);if(composite�0,debug(prime),);prime++)' @ error/fx.c/FxGetSymbol/1831.

Veeeeery interesting!

When I look at the dbg.txt in vim I see:

magick: unable to parse expression `for(prime=2,prime<30,composite=0;for(nn=2,nn<(prime/2+1),if((prime%nn)ù0,composite++,);nn++);if(composite÷0,debug(prime),);prime++)' @ error/fx.c/FxGetSymbol/1831.

Note carefully where the � would be. They are two characters and they are different in the two locations. I think this is very, very odd.

**Reply by rickan54** [reply]
In my case it also looks like IM is getting the correct characters. But then something happens.

**Comment by snibgo** [comment]
In the first case, instead of "==" we get a weird character. In the second, it is at "<=". I suspect something is translating each double-character into the UTF codes for single glyphs.

Okay, next test. How are you running the command? Do you paste something into the console, or do you run a shell script from a file? If you do it from a file, please zip that file and paste a link to it.

If you paste into the console, please paste that into a file, zip it, and paste a link.

**Reply by rickan54** [reply]
For that last test I I copied and pasted from your post.

Since then I've pasted it into vi and two disparate apps and they all look correct I'll even paste it here:

magick.exe -debug command xc: -channel gray -fx " \
  for (prime=2, prime < 30, composite=0; \
    for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++); \
      if (composite <= 0, debug(prime), ); prime++)" null: 2>dbg.txt

Here it is pasted into a file (via mousepad not vi):

[junk.txt](https://github.com/ImageMagick/ImageMagick/files/7333622/junk.txt)

But the test output, dbg.txt, does show that IM got the correct characters at least it regurgitated the what look like the correct characters in the debug output. They're only incorrect in the error output from IM.

**Reply by rickan54** [reply]
The version of IM in my original post was te AppImage executed through a link.

The installed version is:

Version: ImageMagick 6.8.9-9 Q16 x86_64 2020-12-04 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2014 ImageMagick Studio LLC
Features: DPC Modules OpenMP
Delegates: bzlib cairo djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png rsvg tiff wmf x xml zlib

And it has the same defect.

I'm downloading the latest AppImage.

**Reply by rickan54** [reply]
I can't run the latest AppImage.

$ magick
/tmp/.mount_magickNEsgZf/usr/bin/magick: /lib/x86_64-linux-gnu/libz.so.1: version `ZLIB_1.2.9' not found (required by /tmp/.mount_magickNEsgZf/usr/lib/libpng16.so.16)
/tmp/.mount_magickNEsgZf/usr/bin/magick: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /tmp/.mount_magickNEsgZf/usr/lib/libdjvulibre.so.21)
/tmp/.mount_magickNEsgZf/usr/bin/magick: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /tmp/.mount_magickNEsgZf/usr/lib/libIlmImf-2_2.so.22)
/tmp/.mount_magickNEsgZf/usr/bin/magick: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /tmp/.mount_magickNEsgZf/usr/lib/libbsd.so.0)

$ ll /lib/x86_64-linux-gnu/libz.so.1
lrwxrwxrwx 1 root root 13 Jan 21  2020 /lib/x86_64-linux-gnu/libz.so.1 -> libz.so.1.2.8
$ ll /lib/x86_64-linux-gnu/libm.so.6
lrwxrwxrwx 1 root root 12 Apr 21 12:13 /lib/x86_64-linux-gnu/libm.so.6 -> libm-2.23.so
$ ll /lib/x86_64-linux-gnu/libm.so.6
lrwxrwxrwx 1 root root 12 Apr 21 12:13 /lib/x86_64-linux-gnu/libm.so.6 -> libm-2.23.so
$ ll /lib/x86_64-linux-gnu/libc.so.6
lrwxrwxrwx 1 root root 12 Apr 21 12:13 /lib/x86_64-linux-gnu/libc.so.6 -> libc-2.23.so

I'm very disappointed. AppImages are not as self contained as I'd hoped. Not having the necessary libc in the image is a pretty large flaw IMO. But maybe that's because libc too close to the kernal.

Well, I guess my project is not going to happen.

**Comment by snibgo** [comment]
Yeah, that proves it to my satisfaction. You are running IM v7.0.8-0. If you can, it is worth upgrading to the current version and see if that has the problem.

Your version of IM is getting the correct characters. Something inside IM is translating the two-character strings "==" and "<=" to something weird, and that makes "-fx ..." fail.

What in IM is doing that translation? I have no idea.

A developer needs to take a look.

**Reply by rickan54** [reply]
I'm wondering whether it's something Ubuntu/Debian did.

I think upgrading will be a problem for me since I keep my system up-to-date and that's the version I have, i.e. Ubuntu/Debian seem not to have adopted a later version for 16.04. I will not upgrade my OS because of Snap. Ubuntu has gotten vilely stupid about automatic updates with Snap and their policy is completely incompatible with my horrible, tenuous, slow, limited connection to the Net.

Is there an AppImage version of IM?

**Reply by rickan54** [reply]
The answer to my last question is yes and I already have an IM AppImage. I'm going to try it.

**Reply by rickan54** [reply]
I posted replies to your previous post that I intended to, or should have, posted to this post.

**Comment by snibgo** [comment]
Personally, I rarely reply to posts, because then they become tangled. Instead, I do top-level replies. Anyhow...

Sorry, I know nothing about AppImages.

I have just tried the command in a v7.0.8-64 built by me (with Cygwin GNU tools on Windows 8.1). And this gives me errors similar to yours, at the same places.

dbg.txt is:
```
2021-10-13T00:36:47+00:00 0:00.468 0.000u 7.0.8 Command magick[1339]: magick-cli.c/MagickImageCommand/709/Command
  Initializing "magick"
2021-10-13T00:36:47+00:00 0:00.523 0.000u 7.0.8 Command magick[1339]: magick-cli.c/ProcessCommandOptions/388/Command
  at CLI arg 1 - Starting ("-debug")
2021-10-13T00:36:47+00:00 0:00.529 0.000u 7.0.8 Command magick[1339]: operation.c/CLISettingOptionInfo/440/Command
  at CLI arg 1 - Setting Option: -debug "command" "(null)"
2021-10-13T00:36:47+00:00 0:00.551 0.000u 7.0.8 Command magick[1339]: operation.c/CLINoImageOperator/4701/Command
  at CLI arg 3 - NoImage Operator: -read "xc:" ""
2021-10-13T00:36:48+00:00 0:01.409 0.562u 7.0.8 Command magick[1339]: operation.c/CLISimpleOperatorImages/3665/Command
  at CLI arg 4 - Simple Operator: -channel "gray" "(null)"
2021-10-13T00:36:48+00:00 0:01.451 0.562u 7.0.8 Command magick[1339]: operation.c/CLIListOperatorImages/3755/Command
  at CLI arg 6 - List Operator: -fx "   for (prime=2, prime < 30, composite=0;     for (nn=2, nn < (prime/2+1), if ((prime % nn) == 0, composite++, ); nn++);       if (composite <= 0, debug(prime), ); prime++)" "null"
2021-10-13T00:36:48+00:00 0:01.738 0.593u 7.0.8 Command magick[1339]: magick-cli.c/MagickImageCommand/804/Command
  Cleanup "magick"
2021-10-13T00:36:48+00:00 0:01.739 0.593u 7.0.8 Command magick[1339]: magick-cli.c/MagickImageCommand/847/Command
  Exiting "magick"
magick: unable to parse expression `for(prime=2,prime<30,composite=0;for(nn=2,nn<(prime/2+1),if((prime%nn)ù0,composite++,);nn++);if(composite÷0,debug(prime),);prime++)' @ error/fx.c/FxGetSymbol/1868.
```
As with you, the "at CLI arg 6" line is good, but we get a final error line with two bad characters at the same places as you.

In a way, this is good news as it suggests the problem is not specific to your build of IM.

**Comment by snibgo** [comment, marked as answer]
Ahhh, the plot thickens.

The error message with the weird characters is a red herring. In fx.c, IM translates 2-character operators such as "==" into single characters. When it finds and reports an error, it shows the translated string. This doesn't mean the translations are wrong. It simply means there is an error somewhere in the "-fx" expression.

In v7.0.8-4 fx.c, the code that process a "while" loop starts with this:
```
      if (IsFxFunction(expression,"while",5) != MagickFalse)

```
But I can't find the code for "if" or "for".

The changelog.txt for 7.0.9-14 says "Support Fx for() iterator." This implies it wasn't available in previous versions. So that is why it failed for my v7.0.8-4 but succeeded for my 7.1.0-4.

And the source code for v7.1.0-4 _does_ contain code for "for" and "if".

And that is why it fails for your 7.0.8-0.

Problem solved.

**Reply by rickan54** [reply]
My first idea was that the if was the problem! I spent quite a while looking at it and comparing with the examples and was very frustrated I could find no problem with it! Then I saw the �…

Anyway, that does NOT solve my problem, only from your point of view, but maybe the Mint Linux distribution will. I completed my project by writing the equivalent of the -Fx I wanted in Perl and using Get/Setpixel.

Could I have used ?: (ternary conditional) instead of if? Maybe I'll try it and see. Then I can compare the speed with Perl.

**Reply by rickan54** [reply]
?: is parsed in my version of IM.

So now I have other problems with -Fx, mainly the documentation.

**Reply by rickan54** [reply]
I've now read the part of the -Fx documentation which describes why -Fx is NOT the way to do my project anyway and so Get/Setpixel was the way to go and -Fx was a distraction. It's not conclusive but my reading leads me to believe that what I did is not possible using -Fx even the newest version.

What I did was convert each pixel r, g, b to hue, saturation, and lightness and then folded the hue on the yellow-blue axis to the green-cyan side, converted the new h, s, l back to r, g, b and wrote it to the pixel. And since my Perl code which did that did not have to be re-interpreted for each pixel it was probably faster than -Fx which is re-interpreted for each r, g, b of each pixel.

If the above can be done in -Fx I'd be happy to get the instruction if anyone is willing.

**Comment by snibgo** [comment]
You are right, merely finding what causes a problem doesn't solve it.

If the "-fx" interpreter had smarter error messages such as "token 'for' not understood" I would have saved a few hours.

Yes, ternary works. You can test individual features with a simple command:
```
magick xc:sRGB(10%,20%,30%) -fx "u<0.15?0:1" txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (0,65535,65535)  #0000FFFFFFFF  cyan
```

I suspect that fx "while" works for you. That can be used instead of "for".

As a rule, I don't use "-fx" on large images (with millions of pixels) because it is too slow. The expression string is reinterpreted, including error-checking, for every channel in every pixel. I commented above that operators like "==" are translated at a pre-processing stage into single characters. The same could be done for _all_ operators such as "for(" and "if(", and this would improve performance. Even better, pre-processing would build a proper tree structure of expressions, a pseudo-compilation. I think this would make "-fx" usable for mega-pixel images.

IM contains sufficient power that we can do most operations that we need without fx. For example, your "fold the hue on the yellow-blue axis to the green-cyan side" can be done by decomposing hue into two channels, negating one, and re-composing with a mask.

Sometimes this is inconvenient, and a solution in C is easier. I don't use Perl, but I can see that could also work well.
